metals/src/main/scala/scala/meta/internal/metals/FileWatcher.scala

```diff
@@ -37,62 +37,103 @@ final class FileWatcher(
     didChangeWatchedFiles: DirectoryChangeEvent => Unit
 ) extends Cancelable {
 
-  private val executor = Executors.newFixedThreadPool(1)
-  ThreadPools.discardRejectedRunnables("FileWatcher.executor", executor)
-  private var activeWatcher: Option[DirectoryWatcher] = None
-  private var watching: CompletableFuture[Void] = new CompletableFuture()
+  private val directoryExecutor = Executors.newFixedThreadPool(1)
+  private val fileExecutor = Executors.newFixedThreadPool(1)
+
+  ThreadPools.discardRejectedRunnables(
+    "FileWatcher.executor",
+    directoryExecutor
+  )
+  ThreadPools.discardRejectedRunnables("FileWatcher.fileExecutor", fileExecutor)
+
+  private var activeDirectoryWatcher: Option[DirectoryWatcher] = None
+  private var activeFileWatcher: Option[DirectoryWatcher] = None
+
+  private var directoryWatching: CompletableFuture[Void] =
+    new CompletableFuture()
+  private var fileWatching: CompletableFuture[Void] = new CompletableFuture()
 
   override def cancel(): Unit = {
     stopWatching()
-    executor.shutdown()
-    activeWatcher.foreach(_.close())
+    directoryExecutor.shutdown()
+    fileExecutor.shutdown()
+    activeDirectoryWatcher.foreach(_.close())
+    activeFileWatcher.foreach(_.close())
   }
 
   def restart(): Unit = {
-    val directoriesToWatch = new util.ArrayList[Path]()
+    val sourcesDirectoriesToWatch = new util.ArrayList[Path]()
+    val sourcesFilesToWatch = new util.ArrayList[Path]()
     val createdSourceDirectories = new util.ArrayList[AbsolutePath]()
-    def watch(dir: AbsolutePath, isSourceDirectory: Boolean): Unit = {
-      if (!dir.isDirectory) {
-        dir.createDirectories()
-        if (isSourceDirectory) createdSourceDirectories.add(dir)
+    def watch(path: AbsolutePath, isSource: Boolean): Unit = {
+      if (!path.isDirectory && !path.isFile) {
```

Comment:
please, add a comment that we create those paths just to get a notification they don't exist

Reply:
added